### ISSUE TITLE
Sample Packing cause loss 0 and ppl 1 for Qwen35

### ISSUE THREAD

**Issue by maymosido** [axolotl-ai-cloud/axolotl#3453]
### Please check that this issue hasn't been reported before.

- [x] I searched previous [Bug Reports](https://github.com/axolotl-ai-cloud/axolotl/labels/bug) didn't find any similar reports.

### Expected Behavior

I am fine tuning Qwen3.5-9B for machine translation. For that I installed the latest version of Axolotl, Transformres and some other dependencies.

As you will see in the logs. When setting `sample_packing: true`
 and `pad_to_sequence_len: true` The model just stopped learning and the `loss` get to 0 and `ppl` to 1
After some debugging and search, I decided to set these two parameters to `false` and the training started to work as expected 

[deps.txt](https://github.com/user-attachments/files/25741397/deps.txt)
[qwen35_9B_no_sample_packing.log](https://github.com/user-attachments/files/25741399/qwen35_9B_no_sample_packing.log)
[qwen35_9B_with_sample_packing_failed.log](https://github.com/user-attachments/files/25741398/qwen35_9B_with_sample_packing_failed.log)

### Current behaviour

Training to run without issues.

### Steps to reproduce

This how I run the training:

```
source .venv/bin/activate

export PYTORCH_CUDA_ALLOC_CONF=expandable_segments:True
export WANDB_API_KEY=$(cat ~/.wandb_key)
export HF_TOKEN=$(cat ~/.cache/hf_key)
export ACCELERATE_MIXED_PRECISION=bf16
export CUDA_LAUNCH_BLOCKING=1
export TORCH_NCCL_ASYNC_ERROR_HANDLING=1
export TORCH_NCCL_BLOCKING_WAIT=1
export CUDA_HOME=/usr/local/cuda-12.4
export TORCH_CUDA_ARCH_LIST="8.0;8.6;9.0"
export AXOLOTL_DO_NOT_TRACK=1

DATE=$(date +%Y%m%d_%H%M)
LOG_DIR=logs

LOG_FILE=$LOG_DIR/qwen35_9B_10M_$DATE.log
#############################
# Train
#############################

CUDA_VISIBLE_DEVICES=0,1,2,3,4,5,6,7 accelerate launch \
    --main_process_port 29502 \
    --num_processes 8 \
    --num_machines 1 \
    --mixed_precision bf16 \
    --dynamo_backend no \
    -m axolotl.cli.train configs/qwen35.yaml \
    2>&1 | tee $LOG_FILE
```

### Config yaml

```yaml
base_model: Qwen/Qwen3.5-9B
model_type: AutoModelForCausalLM
tokenizer_type: AutoTokenizer

skip_prepare_dataset: false
dataset_prepared_path: /datasets/processed/qwen35-9B/10M/

eager_attention: false
load_in_8bit: false
load_in_4bit: false
strict: false

chat_template: tokenizer_default
  
datasets:
  - path: /datasets/train/balanced_dist_merged_src_langs/train_10M/
    type: chat_template
    field_messages: conversations
    drop_system_message: false
    message_property_mappings:
      role: role
      content: content
    roles_to_train: ["assistant"]

val_set_size: 0
output_dir: /experiments/qwen35_9B_custom_template_10M/

use_tensorboard: true

sequence_len: 256
sample_packing: false
pad_to_sequence_len: false

adapter:
lora_model_dir:
lora_r:
lora_alpha:
lora_dropout:
lora_target_linear:
lora_fan_in_fan_out:


bf16: true                     
fp16: false
trust_remote_code: false


# tie_word_embeddings: false

gradient_accumulation_steps: 4
micro_batch_size: 16
eval_batch_size: 16
num_epochs: 3
optimizer: adamw_torch
lr_scheduler: cosine
learning_rate: 7e-6

gradient_checkpointing: true
early_stopping_patience:
local_rank:
logging_steps: 1
xformers_attention: false
flash_attention: true
# attn_implementation: sdpa
attn_implementation: flash_attention_2
flash_attn_cross_entropy: false
flash_attn_rms_norm: true
flash_attn_fuse_qkv: false
flash_attn_fuse_mlp: false

warmup_steps: 125
# warmup_ratio: 0.03
eval_steps:
eval_sample_packing: False
eval_table_size:
save_strategy: "epoch"
saves_per_epoch: 1
save_total_limit: 8
save_only_model: false

debug:
deepspeed: zero2_new.json
weight_decay: 0.01
fsdp:
fsdp_config:

tokens:
special_tokens:
  bos_token: "<|im_start|>"
  eos_token: "<|im_end|>"
  pad_token: "<|im_end|>"
```

### Possible solution

In the config you see both `sample_packing` and `pad_to_sequence_len` are set to false. But for different experiments, I changed these values.

There is an old similar issue, but I was recommend by a developer to open a new one.

### Which Operating Systems are you using?

- [x] Linux
- [ ] macOS
- [ ] Windows

### Python Version

3.11

### axolotl branch-commit

77828d3559926c78d54506d65338cada68544d8c

### Acknowledgements

- [x] My issue title is concise, descriptive, and in title casing.
- [x] I have searched the existing issues to make sure this bug has not been reported yet.
- [x] I am using the latest version of axolotl.
- [x] I have provided enough information for the maintainers to reproduce and diagnose the issue.

**Comment by maymosido** [comment]
Sometimes the when setting the `sample_packing: true` and `pad_to_sequence_len: true` the experiment ends with error. See the attached logs.

[qwen35_9B_packing_failure_with_error.log](https://github.com/user-attachments/files/25760168/qwen35_9B_packing_failure_with_error.log)

**Comment by NanoCode012** [comment]
@maymosido , thanks for opening this. I lost the link to the old issue, so could you just add a link somewhere to link this? 

Next, could we try to simplify your config? Can you start with one of our example configs and then start enabling configs you have in yours to narrow down the issue? I've been running off main with some of our main configs and did not show this behavior. (Preferably, use the public datasets already in the config).

For the model arch qwen 35 specifically, please see #3442

**Comment by maymosido** [comment]
The old issue link:
https://github.com/axolotl-ai-cloud/axolotl/issues/1706#issuecomment-3996526500

Here is my config which I do not think is very different from the examples in Axolotl repo:

```
base_model: Qwen/Qwen3.5-9B
model_type: AutoModelForCausalLM
tokenizer_type: AutoTokenizer

skip_prepare_dataset: false
dataset_prepared_path: /datasets/processed/qwen35-9B/10M/

chat_template: tokenizer_default
  
datasets:
  - path: /datasets/train/balanced_dist_merged_src_langs/train_10M/
    type: chat_template
    field_messages: conversations
    drop_system_message: false
    message_property_mappings:
      role: role
      content: content
    roles_to_train: ["assistant"]

val_set_size: 0
output_dir: /experiments/qwen35_9B_custom_template_10M/

use_tensorboard: true

sequence_len: 8192
sample_packing: true
pad_to_sequence_len: true


load_in_8bit: false
load_in_4bit: false
strict: false

adapter:
lora_model_dir:
lora_r:
lora_alpha:
lora_dropout:
lora_target_linear:
lora_fan_in_fan_out:

bf16: true
fp16: false


gradient_accumulation_steps: 16
micro_batch_size: 1
eval_batch_size: 1
num_epochs: 3
optimizer: adamw_torch
lr_scheduler: cosine
learning_rate: 7e-6

gradient_checkpointing: true
early_stopping_patience:
local_rank:
logging_steps: 1
xformers_attention: false
eager_attention: false
flash_attention: true
# attn_implementation: sdpa
attn_implementation: flash_attention_2
flash_attn_cross_entropy: false
flash_attn_rms_norm: true
flash_attn_fuse_qkv: false
flash_attn_fuse_mlp: false

warmup_steps: 125
eval_steps:
eval_sample_packing: False
eval_table_size:
save_strategy: "epoch"
saves_per_epoch: 1
save_total_limit: 8
save_only_model: false

debug:
deepspeed: /deepspeed_configs/zero2_new.json
weight_decay: 0.01
fsdp:
fsdp_config:

tokens:
special_tokens:
  bos_token: "<|im_start|>"
  eos_token: "<|im_end|>"
  pad_token: "<|im_end|>"
```

**Comment by maymosido** [comment]
The issue seems to be related to flash attention. Instead of using `attn_implementation: flash_attention_2` , I used `attn_implementation: sdpa` and the training is working fine and as expected.

```
attn_implementation: sdpa
```

**Comment by NanoCode012** [comment]
> The issue seems to be related to flash attention. Instead of using `attn_implementation: flash_attention_2` , I used `attn_implementation: sdpa` and the training is working fine and as expected.
> 
> ```
> attn_implementation: sdpa
> ```

Could you try, 
```diff
flash_attention: true
-# attn_implementation: sdpa
-attn_implementation: flash_attention_2
-flash_attn_cross_entropy: false
-flash_attn_rms_norm: true
-flash_attn_fuse_qkv: false
-flash_attn_fuse_mlp: false
```

Let's just try base FA2?

**Comment by maymosido** [comment]
I have already tried with these settings.
```
flash_attention: true
-# attn_implementation: sdpa
-attn_implementation: flash_attention_2
-flash_attn_cross_entropy: false
-flash_attn_rms_norm: true
-flash_attn_fuse_qkv: false
-flash_attn_fuse_mlp: false
```
If you check my original config above, you will see they are set exactly as you asked me to.

**Comment by NanoCode012** [comment]
I noticed that you set both 
```
flash_attention: true
attn_implementation: flash_attention_2
```

I'm just guessing they may be potentially conflicting. I'm not sure we had a check against this. Please only keep the first line and see

**Comment by maymosido** [comment]
I tested with only keeping  `flash_attention: true` and still get the same results.

```
'loss': '0', 'grad_norm': '1.414', 'learning_rate': '1.064e-06', 'ppl': '1'
```

**Comment by NanoCode012** [comment]
Thanks for giving that a try. I would say, for qwen35 (as shown in your example), please use the linked PR #3442 . Packing does not work with original transformers source, we have to patch it, hence that linked PR. We did the same for qwen3_next.

I looked over your config and it really does look standard, so I'm suspecting the above.

If you're open to test, maybe try a simple `base_model: meta-llama/Llama-3.2-1B` to verify it's not from the arch specifically.

**Comment by maymosido** [comment]
For now, I am training with `attn_implementation: sdpa` and once it is done, I will test with your suggestion. 
Thanks for your quick reply.

**Comment by NanoCode012** [comment]
> For now, I am training with `attn_implementation: sdpa` and once it is done, I will test with your suggestion. Thanks for your quick reply.

How did it go?

**Comment by maymosido** [comment]
> > For now, I am training with `attn_implementation: sdpa` and once it is done, I will test with your suggestion. Thanks for your quick reply.
> 
> How did it go?

Currently I am also fine tuning the Qwen3.5-27B, this will finish by next week and once done I can test the patch you suggested

**Comment by jjbuschhoff** [comment]
I had a similar issue once where in some of my sequences, all tokens' attention was masked. This caused zero-losses with FA2 but no problem with SDPA. It might be a case that packing and truncating causes some issues on your data specifically.

**Comment by NanoCode012** [comment]
> I had a similar issue once where in some of my sequences, all tokens' attention was masked. This caused zero-losses with FA2 but no problem with SDPA. It might be a case that packing and truncating causes some issues on your data specifically.

Do you perhaps have a repro for that? Could it be a prompter/dataset specific issue?

**Comment by jjbuschhoff** [comment]
> > I had a similar issue once where in some of my sequences, all tokens' attention was masked. This caused zero-losses with FA2 but no problem with SDPA. It might be a case that packing and truncating causes some issues on your data specifically.
> 
> Do you perhaps have a repro for that? Could it be a prompter/dataset specific issue?

Sorry, that was a while ago and not using axolotl (just plain `transformers` with some buggy handcrafted attention masking).

**Comment by onder-prosus** [comment]
I hit the same loss=0 issue with Qwen3.5 + packing + flash_attention_2. Root cause is that `cu_seqlens` never reaches fla's `chunk_gated_delta_rule` and `seq_idx` is hardcoded to `None` in the conv path — so recurrent state leaks across packed sequence boundaries.

Got it working with a monkey-patch that passes `cu_seqlens`/`seq_idx` to the fla and conv kernels, plus a fix for `_is_packed_sequence()` misinterpreting Qwen3.5's 3D position_ids (#44910 in transformers).

More details in huggingface/transformers#44717.

Tested on 0.8B, 2B, 4B, all pass with gradient checkpointing.

**Comment by NanoCode012** [comment]
Thanks @onder-prosus , we already solved this in this repo a while back. https://github.com/axolotl-ai-cloud/axolotl/blob/main/src/axolotl/monkeypatch/models/qwen3_5/modeling.py 

I also mentioned it in an earlier comment https://github.com/axolotl-ai-cloud/axolotl/issues/3453#issuecomment-4003323255 

I'll close this issue then. If OP still has the problem, please re-open.
